File: 110/balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
    let balanced = true;

    function dfs(root) {
        if (!root)
            return;
        let l_height = root.left ? dfs(root.left) : 0;
        let r_height = root.right ? dfs(root.right) : 0;

        if (Math.abs(l_height - r_height) > 1)
            balanced = false;

        return Math.max(l_height, r_height) + 1;
    }

    dfs(root);
    return balanced;
};

File: 110/BalancedBinaryTree.scala
object BalancedBinaryTree {
  def isBalanced(root: TreeNode): Boolean = {
    def isBalanced(root: TreeNode): (Boolean, Int) =
      if (root == null)
        (true, 0)
      else {
        val (lb, lh) = isBalanced(root.left)
        val (rb, rh) = isBalanced(root.right)
        if (!lb || !rb)
          (false, 0)
        else if ((lh - rh).abs > 1)
          (false, 0)
        else
          (true, (lh max rh) + 1)
      }

    isBalanced(root)._1
  }

  class TreeNode(var _value: Int = 0) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }
}
